### PR TITLE
Add `Jsonify` type

### PR DESCRIPTION
**Pull request by darcyparker**
I would like to add the very useful ~`IsJsonable`~ `Jsonify` utility.  The credit for this type alias goes to @grant-dennison and other participants in this issue: https://github.com/Microsoft/TypeScript/issues/1897#issuecomment-710744173 

I have simply refactored it slightly to satisfy `xo` rules, added some remarks to the tsdoc, and refactored the tests into `tsd`.

I also updated `JsonValue` in `basic.d.ts`; which was annotated as TODO.  (There are other existing TODO comments that are throwing `no-warning-comments` that I did not add.)

I am sure https://github.com/Microsoft/TypeScript/issues/1897 will be addressed eventually.  But in the meantime, having a recursive `JsonValue` type alias (without interface) and this ~`IsJsonable`~ `Jsonify` utility are very helpful.

### CODE REVIEW

**Comment by darcyparker** [comment]
Typescript 3.4 support will need to be dropped because recursive type aliases aren't allowed in it.  Is that acceptable?

File: source/is-jsonable.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                    ? {
+                        // Iterate over keys in object case
+                        [P in keyof T]:
+                            // Recursive call for children
+                            IsJsonable<T[P]>
+                    }
+                    // Otherwise any other non-object no bueno
+                    : never
+        // Otherwise non-jsonable type union was found not empty
+        : never;
```
Comment (sindresorhus) [inline source/is-jsonable.ts:49]: Use tab-indentation.

File: source/is-jsonable.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import {JsonPrimitive} from './basic';
+/**
+Transforms type T into a type that can be structurally matched to `JsonValue` type above
+
+@remarks:
+
+An interface cannot be structurally compared to `JsonValue` because an interface can be re-opened
+to add properties that may not be satisfy `JsonValue`.
```
Comment (sindresorhus) [inline source/is-jsonable.ts:8]: Don't hard-wrap.

File: readme.md [hunk 3]
```diff
@@ -120,6 +120,7 @@ Click the type names for complete docs.
 - [`SetReturnType`](source/set-return-type.d.ts) - Create a function type with a return type of your choice and the same parameters as the given function type.
 - [`Asyncify`](source/asyncify.d.ts) - Create an async version of the given function type.
 - [`Simplify`](source/simplify.d.ts) - Flatten the type output to improve type hints shown in editors.
+- [`IsJsonable`](source/is-jsonable.d.ts) - Transforms type T into a type that can be structurally matched to [`JsonValue`](source/basic.d.ts) type alias
```
Comment (sindresorhus) [inline readme.md:123]: Needs a simpler description.

**Comment by sindresorhus** [comment]
I'm confused by the name. Why does a type that starts with `Is` "transform"?

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/basic.d.ts [hunk 4]
```diff
@@ -20,13 +20,20 @@ Matches a JSON array.
...
-// TODO: Make it this when targeting TypeScript 4.1:
-// export type JsonArray = JsonValue[];
+export type JsonArray = JsonValue[];
+
+/**
+Matches any valid JSON primitive value.
+
+@category Basic
+*/
+export type JsonPrimitive = string | number | boolean | null;
```
Comment (sindresorhus) [inline source/basic.d.ts:30]: This will have to be documented in the readme too.

File: source/basic.d.ts [hunk 5]
```diff
@@ -20,13 +20,20 @@ Matches a JSON array.
...
 
 /**
 Matches any valid JSON value.
 
+@remarks: Also see ./IsJsonable
```
Comment (sindresorhus) [inline source/basic.d.ts:35]:
```suggestion
@remarks: Also see ./IsJsonable
```

I don't think it's correct to use `:`

File: source/is-jsonable.ts (hunk 1 above)
Comment (darcyparker) [reply source/is-jsonable.ts]: Sorry. Fixed in 14a910c

File: source/is-jsonable.ts (hunk 2 above)
Comment (darcyparker) [reply source/is-jsonable.ts]: Fixed in 14a910c

File: source/basic.d.ts (hunk 4 above)
Comment (darcyparker) [reply source/basic.d.ts:30]: See 03a7407

File: source/basic.d.ts (hunk 5 above)
Comment (darcyparker) [reply source/basic.d.ts]: I removed the `:`

File: readme.md (hunk 3 above)
Comment (darcyparker) [reply readme.md]: How is "Transforms a type to a type that is assignable to `JsonValue` type". (This is what I wrote in most recent commits).

I am open to discussion.  It is kind of like a mapped type, but it does so recursively. And it is a kind of identity for primitives and primitive properties.  So I chose 'transforms'.   And I think it is simpler to say the new type is "assignable".

**Comment by darcyparker** [comment]
> I'm confused by the name. Why does a type that starts with `Is` "transform"? That's usually used for things that return a boolean. I would think `Jsonify` or something would be more appropriate.

I agree. I was using the original name. (Which I think was probable named with `Is` because it can also be thought of as a kind of assertion to.  The person is declaring this interface or type that contains a reference to an interface is a `JsonValue`.) 

But I see your point. It is renamed.

File: readme.md (hunk 3 above)
Comment (sindresorhus) [reply readme.md]: > Transforms a type to a type that is assignable to JsonValue type

Sounds good, but `transforms` => `transform`.

File: readme.md [hunk 6]
```diff
@@ -120,6 +121,7 @@ Click the type names for complete docs.
 - [`SetReturnType`](source/set-return-type.d.ts) - Create a function type with a return type of your choice and the same parameters as the given function type.
 - [`Asyncify`](source/asyncify.d.ts) - Create an async version of the given function type.
 - [`Simplify`](source/simplify.d.ts) - Flatten the type output to improve type hints shown in editors.
+- [`Jsonify`](source/jsonify.d.ts) - Transforms a type to a type that is assignable to `JsonValue` type
```
Comment (sindresorhus) [inline readme.md:124]:
```suggestion
- [`Jsonify`](source/jsonify.d.ts) - Transform a type to a type that is assignable to the `JsonValue` type
```

**Comment by sindresorhus** [comment]
You force-pushed over my small changes in https://github.com/sindresorhus/type-fest/commit/f096e965070644b665fc334e3a7c3ab3c5cb1957

**Comment by darcyparker** [comment]
> You force-pushed over my small changes in [f096e96](https://github.com/sindresorhus/type-fest/commit/f096e965070644b665fc334e3a7c3ab3c5cb1957)

I just noticed that. I am pretty sure I did not force pushed. I made the change in my fork's branch before I received f096e96.  When I pushed, I then saw an email about f096e965070644b665fc334e3a7c3ab3c5cb1957 in https://github.com/sindresorhus/type-fest/.  I was going to merge it in, but with the rename of IsJsonable to Jsonify, I incorporated the change in 343e89b.  I can rework this to put f096e96 in if you'd like. But I would have to force push or close the pull request and start with new branch.

File: readme.md [hunk 7]
```diff
@@ -120,6 +121,7 @@ Click the type names for complete docs.
 - [`SetReturnType`](source/set-return-type.d.ts) - Create a function type with a return type of your choice and the same parameters as the given function type.
 - [`Asyncify`](source/asyncify.d.ts) - Create an async version of the given function type.
 - [`Simplify`](source/simplify.d.ts) - Flatten the type output to improve type hints shown in editors.
+- [`Jsonify`](source/jsonify.d.ts) - Transform a type to a type that is assignable to the `JsonValue` type
```
Comment (sindresorhus) [inline readme.md:124]:
```suggestion
- [`Jsonify`](source/jsonify.d.ts) - Transform a type to one that is assignable to the `JsonValue` type
```

How about this? Felt like too many "type" words.

File: source/basic.d.ts [hunk 8]
```diff
@@ -20,13 +20,20 @@ Matches a JSON array.
...
+Matches any valid JSON primitive value.
+
+@category Basic
+*/
+export type JsonPrimitive = string | number | boolean | null;
 
 /**
 Matches any valid JSON value.
 
+@remarks Also see ./Jsonify
```
Comment (sindresorhus) [inline source/basic.d.ts:35]: Use the `@see` keyword (make sure you use it correctly)

File: source/jsonify.d.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import {JsonPrimitive} from './basic';
+
+type NotJsonable = ((...args: any[]) => any) | undefined;
```
Comment (sindresorhus) [inline source/jsonify.d.ts:3]:
```suggestion
type NotJsonable = ((...args: any[]) => unknown) | undefined;
```

**Comment by sindresorhus** [comment]
You didn't actually update https://github.com/sindresorhus/type-fest/pull/244/files#r682193697

File: readme.md (hunk 7 above)
Comment (sindresorhus) [reply readme.md:124]: The description should also end in a `.`

File: source/jsonify.d.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+Transform a type to a type that is assignable to the `JsonValue` type.
+
+@remarks
+
+An interface cannot be structurally compared to `JsonValue` because an interface can be re-opened to add properties that may not be satisfy `JsonValue`.
+
+Without intending to pick on @types/geojson, sometimes there are Json type definitions like GeoJson's [`Point` interface(https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/geojson/index.d.ts#L86-L89 that) that are clearly Json and won't be extended.  If a method requires a `JsonValue` argument such as in `const someMethodOnJson = <T extends JsonValue>(data: T) => void`, and you pass a value of an interface type like `GeoJSON.Point`, then there will be a type error.  To solve this problem, you can assert the interface type is JSON like this `Jsonify<GeoJSON.Point>`.  `Jsonify<T>` is also useful for types that have nested interface types.
+
```
Comment (sindresorhus) [inline source/jsonify.d.ts:13]: > Example code block with a realistic example. - https://github.com/sindresorhus/type-fest/blob/main/.github/contributing.md

File: source/jsonify.d.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+	? {
+		// Iterate over keys in object case
+		[P in keyof T]:
+		// Recursive call for children
+		Jsonify<T[P]>
+	}
+	// Otherwise any other non-object no bueno
+	: never
+	// Otherwise non-jsonable type union was found not empty
+	: never;
```
Comment (sindresorhus) [inline source/jsonify.d.ts:45]: The conditionals here are really hard to follow. Use parens for nested conditions and also increase the indentation for those.

File: source/jsonify.d.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+	// Otherwise check if object
+	: T extends object
+	// It's an object
+	? {
+		// Iterate over keys in object case
+		[P in keyof T]:
+		// Recursive call for children
+		Jsonify<T[P]>
+	}
+	// Otherwise any other non-object no bueno
```
Comment (sindresorhus) [inline source/jsonify.d.ts:42]: English-only.

File: source/jsonify.d.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+type NotJsonable = ((...args: any[]) => any) | undefined;
+
+/**
+Transform a type to a type that is assignable to the `JsonValue` type.
+
+@remarks
+
+An interface cannot be structurally compared to `JsonValue` because an interface can be re-opened to add properties that may not be satisfy `JsonValue`.
+
+Without intending to pick on @types/geojson, sometimes there are Json type definitions like GeoJson's [`Point` interface(https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/geojson/index.d.ts#L86-L89 that) that are clearly Json and won't be extended.  If a method requires a `JsonValue` argument such as in `const someMethodOnJson = <T extends JsonValue>(data: T) => void`, and you pass a value of an interface type like `GeoJSON.Point`, then there will be a type error.  To solve this problem, you can assert the interface type is JSON like this `Jsonify<GeoJSON.Point>`.  `Jsonify<T>` is also useful for types that have nested interface types.
```
Comment (sindresorhus) [inline source/jsonify.d.ts:12]: Json => JSON

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: readme.md (hunk 7 above)
Comment (darcyparker) [reply readme.md]: See 0f760c94deb510f0829ec4dd90e3adb0909c8356

File: source/basic.d.ts (hunk 8 above)
Comment (darcyparker) [reply source/basic.d.ts]: 6c6d324f787855560d67590d48c33befdd64efb4

File: source/jsonify.d.ts (hunk 10 above)
Comment (darcyparker) [reply source/jsonify.d.ts:13]: How is a548bb38cc6d0c713de60613dd17650c6df797f1?

File: source/jsonify.d.ts (hunk 12 above)
Comment (darcyparker) [reply source/jsonify.d.ts]: Fixed in a548bb3

File: source/jsonify.d.ts (hunk 11 above)
Comment (darcyparker) [reply source/jsonify.d.ts]: Does 6ddf972 make it easier to follow? I adjusted the comment positions and used prettier to format it.

File: source/basic.d.ts [hunk 14]
```diff
@@ -32,7 +32,7 @@ export type JsonPrimitive = string | number | boolean | null;
 /**
 Matches any valid JSON value.
 
-@remarks Also see ./Jsonify
+@see Also see ./Jsonify
```
Comment (sindresorhus) [inline source/basic.d.ts:35]: This is not correct usage of `@see`. I specifically asked you to make sure you used it correctly.

File: source/jsonify.d.ts [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+/**
+Transform a type to a type that is assignable to the `JsonValue` type.
+
+@remarks
+
+An interface cannot be structurally compared to `JsonValue` because an interface can be re-opened to add properties that may not be satisfy `JsonValue`.
+
+Example:
+
+interface Geometry {
```
Comment (sindresorhus) [inline source/jsonify.d.ts:14]: The code should be wrapped in a proper code block. See any of the other types in here.

File: source/jsonify.d.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+	// Note: use of tuples in this first condition side-steps distributive conditional types
+	// (see https://github.com/microsoft/TypeScript/issues/29368#issuecomment-453529532)
+	[Extract<T, NotJsonable>] extends [never]
+		? T extends JsonPrimitive
+			? T // Primitive is acceptable
+			: T extends Array<infer U>
+			? Array<Jsonify<U>> // It's an array: recursive call for its children
+			: T extends object
+			? { [P in keyof T]: Jsonify<T[P]> } // It's an object: recursive call for its children
+			: never // Otherwise any other non-object is removed
```
Comment (sindresorhus) [inline source/jsonify.d.ts:53]: The nested conditionals here should be further indented.

File: source/basic.d.ts (hunk 14 above)
Comment (darcyparker) [reply source/basic.d.ts:35]: I added 8bf303296bcbd0bfb8291d0e219c0602d017f5f8

This is similar to https://github.com/sindresorhus/type-fest/blob/6b8bee7d0df2bfce9cab4ae59bb6b14f81d7a633/source/entries.d.ts#L13, but different than others like https://github.com/sindresorhus/type-fest/blob/6b8bee7d0df2bfce9cab4ae59bb6b14f81d7a633/ts41/delimiter-case.d.ts#L25

File: source/jsonify.d.ts (hunk 15 above)
Comment (darcyparker) [reply source/jsonify.d.ts:14]: See 903f94c72d6892788018fc14ad50cece7c551486

File: source/jsonify.d.ts (hunk 16 above)
Comment (darcyparker) [reply source/jsonify.d.ts]: 468e8e367115612754575f7b04ed4e3ae6963718 adds more indentation.

There are many styles of indentation like https://github.com/sindresorhus/type-fest/blob/main/source/partial-deep.d.ts#L33-L47 and https://github.com/sindresorhus/type-fest/blob/main/ts41/camel-case.d.ts#L11-L18.  I presume the latter is preferred.

**Comment by sindresorhus** [comment]
Can you fix the merge conflict? `base.d.ts` was renamed to `index.d.ts`.

**Comment by darcyparker** [comment]
> Can you fix the merge conflict? `base.d.ts` was renamed to `index.d.ts`.

Done.
